Prepare for a stable release

weierophinney opened the pull request:
This patch accomplishes two things:

- It updates the zend-expressive-authentication requirement to `^1.0`. All tests continue to pass as-is.
- It fleshes out the documentation to detail how to configure the package, add a login handler, and add authentication to an existing application (as it requires both the session and authentication middleware).

Fixes #17
Supercedes #18

jonsa committed:
Handle iterable returned from getRoles in UserInterface

weierophinney committed:
Update to zend-expressive-authentication 1.0.0

We have a stable version of the APIs now, so we can safely upgrade.

All tests continue to run.

weierophinney committed:
Full documentation for the package

This patch adds full documentation for the package, detailing:

- configuration
- creating a login handler
- adding authentication to your application

weierophinney commented:
@ezimuel Could you review the documentation to make sure it's correct? I wrote a demo application, but did not have a user repository to test against and validate the approach.

jonsa commented:
@weierophinney As far as I can tell this won't work as is. The `UserInterface::getRoles()` method now returns an `iterable` which includes `Generators`. In my pull request I tried to handle that.

Some nitpicking: The form example in the documentation should include the name attribute on the fields otherwise nothing is posted.

froschdesign commented on mkdocs.yml:
```diff
@@ -2,8 +2,14 @@ docs_dir: docs/book
 site_description: 'Username/password, session-backed authentication adapter for zend-expressive-authentication.'
 repo_url: 'https://github.com/zendframework/zend-expressive-authentication-session'
-copyright: 'Copyright (c) 2017 <a href="http://www.zend.com/">Zend Technologies USA Inc.</a>'
+copyright: 'Copyright (c) 2017-2018 <a href="http://www.zend.com/">Zend Technologies USA Inc.</a>'
```
Can be removed, because this line is no longer used by the documentation theme.

froschdesign commented on docs/book/v1/login-handler.md:
````diff
@@ -0,0 +1,185 @@
+`config/routes.php` file, and edit it to add the following within its callback:
+
+```php
+    $app->get(
````
The indention is not needed for this example.

froschdesign commented on docs/book/v1/login-handler.md:
```diff
@@ -0,0 +1,185 @@
+        <div class="col-sm"><form action="<?= $action ?>" method="post">
+            <div class="form-group">
+                <label for="username">Username</label>
+                <input type="text" class="form-control" id="username" placeholder="Enter username">
```
`name` attribute is missing.

froschdesign commented on docs/book/v1/login-handler.md:
```diff
@@ -0,0 +1,185 @@
+            </div>
+            <div class="form-group">
+                <label for="password">Password</label>
+                <input type="password" class="form-control" id="password" placeholder="Password">
```
`name` attribute is missing.

froschdesign commented on docs/book/v1/config.md:
````diff
@@ -0,0 +1,65 @@
+
+
+```php
+<?php
````
If the code example does not present template code, this line should be omitted.

froschdesign commented on mkdocs.yml:
```diff
@@ -2,8 +2,14 @@ docs_dir: docs/book
 site_dir: docs/html
 pages:
     - index.md
```
Can you add "Home" (`- Home: index.md`) here? I think, then the entry in the subnavigation is skipped.

froschdesign reviewed (changes requested):
Only litte changes necessary.

I like the example with the login handler. 👍

jonsa commented:
I played around with the documentation example a little bit more and I cannot get it to work. Unless I'm missing something obvious I have to enable POST on all my routes since the LoginHandler set the form action to be that of `$redirect`. So when I send the login form I get a 405 (Method not allowed) when I successfully login.

weierophinney commented:
> As far as I can tell this won't work as is. The `UserInterface::getRoles()` method now returns an `iterable` which includes `Generators`. In my pull request I tried to handle that.

When the tests passed as-is, I missed this. But you're right - when we serialize the session, we need to grab all of them. I'll merge your patch into my branch so we get that functionality in place.

weierophinney commented:
> I played around with the documentation example a little bit more and I cannot get it to work. Unless I'm missing something obvious I have to enable POST on all my routes since the LoginHandler set the form action to be that of `$redirect`. So when I send the login form I get a 405 (Method not allowed) when I successfully login.

I had a nagging suspicion that may be the case. I'll post some updates shortly to see if I can resolve this via the `LoginHandler`.

weierophinney committed:
Incorporated feedback from @froschdesign

weierophinney committed:
Fix workflow for LoginHandler

As pointed out by @jonsa, the original workflow did not work, because it
required that any endpoint requiring authentication would respond to
`POST` requests. Since this is almost certainly not what is expected, I
needed to refactor how to handle requests where credentials are
submitted.

The new workflow has the handler respond to two different routes. The
first, a `GET` request, composes only the session middleware and
handler. The second, a `POST` request, composes the session middleware,
authentication middleware, and the handler. This allows successful
credentials to lead back to the handler, which can then redirect to the
originally requested page.

weierophinney commented:
@jonsa I've rewritten the page covering the login handler. Can you review and test to ensure it works as expected, please?

weierophinney commented:
@froschdesign I've incorporated all feedback you presented at this time.

jonsa commented:
@weierophinney Your solution looks more or less the same as what I've come up with so far. It has one problem though. When submitting the login form with invalid credentials `AuthenticationMiddleware` will redirect to the login page and you loose any knowledge that a sign in has occurred and cannot give the user any feedback.

weierophinney commented:
> It has one problem though. When submitting the login form with invalid credentials `AuthenticationMiddleware` will redirect to the login page and you loose any knowledge that a sign in has occurred and cannot give the user any feedback.

This is due to how the adapter works, as the `unauthenticatedResponse()` method performs a redirect.

The only way we could handle that differently is to have middleware intercepting before the `AuthenticationMiddleware` that uses something like zend-expressive-flash to store the previous values and/or a message indicating invalid credentials when it detects a 401 response to the login page... and, frankly, that seems far too complicated.

Alternately, I could modify the `LoginHandler` to also accept the `PhpSession` adapter and perform the authentication attempt itself; that would allow it to re-display the login form for an invalid attempt.

That might actually be a better approach in the end. I'm going to create a patch that does this, and will ping you when it's pushed.

jonsa commented:
My solution at the moment works the other way around. I'm wrapping `PhpSession` with a "smarter" adapter that takes the `LoginHandler` as an argument and internally takes care of handling the store previous url and redirect back on successful login. This makes defining the `/login` route redundant.

The gist of it is that I return `null` in `authenticate` if I'm on the configured login path and redirects back in `unauthorizedResponse` if the session contains `UserInterface`.

weierophinney committed:
Refactor LoginHandler to process credentials

This patch refactors the LoginHandler documentation page as follows:

- I moved creation of the handler to be the first step, mimicing the
  steps in the introduction.
- I rewrote the handler to accept the authentication adapter as a
  constructor argument. During `POST` requests, it uses the adapter
  to validate the credentials. On success, it redirects to the
  originally requested page, and on failure it re-displays the form with
  an error message.
- I updated the template to display an error message when present.
- I updated routing to use `route()` and specify both `GET` and `POST`
  as valid methods. The pipeline now only uses the session middleware
  and login handler.

weierophinney commented:
@jonsa I've updated the docs and pushed just now. We can also provide your approach as an alternative if you'd like; they're both valid, and it might be interesting to demonstrate each.

weierophinney committed:
Remove documentation from README file

- Redundant.

weierophinney committed:
Removes obsolete `@todo` annotation

weierophinney committed:
Final edits to documentation

weierophinney committed:
Adds CHANGELOG entry for #19

weierophinney commented:
@jonsa Definitely send in a PR detailing your approach; we can merge that at any time, and it will immediately publish to the documentation.

icetee commented:
I think it is useful information to overwrite the default `username` field, for example `email`. I had to look at the code all the way, to understand.

```
'username' => null, // provide a custom field name for the username
'password' => null, // provide a custom field name for the password
```


PS: Otherwise more beautiful property name `identity` and default `username`. See `PdoDatabase`.

weierophinney commented:
@icetee Please don't comment on an already merged request. Add a feature request or pull request, but don't comment on something already complete.
